crates/iota-transaction-builder/src/builder.rs

```diff
@@ -0,0 +1,1070 @@
+// Copyright 2024 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+//! Builder for Programmable Transactions.
+
+use core::marker::PhantomData;
+use std::collections::HashMap;
+
+use iota_crypto::IotaSigner;
+use iota_graphql_client::{Client, DryRunResult};
+use iota_types::{
+    Address, Argument, Command, GasPayment, Identifier, Input, MakeMoveVector, MergeCoins,
+    MoveCall, ObjectId, ObjectReference, Owner, Publish, SplitCoins, Transaction,
+    TransactionEffects, TransactionExpiration, TransferObjects, TypeTag, Upgrade,
+};
+use serde::Serialize;
+
+use crate::{
+    error::Error,
+    publish_type::PublishType,
+    types::{MoveParam, MoveType, MoveTypes, ParamType},
+};
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+enum UnresolvedInput {
+    /// A move object that is either immutable or address owned
+    ImmutableOrOwned(ObjectId),
+    /// A move object whose owner is "Shared"
+    Shared {
+        object_id: ObjectId,
+        /// Controls whether the caller asks for a mutable reference to the
+        /// shared object.
+        mutable: bool,
+    },
+    /// A move object that is attempted to be received in this transaction.
+    Receiving(ObjectId),
+    Resolved(Input),
+}
+
+/// A builder for a programmable transaction which provides a better API vs.
+/// IOTAs [`ProgrammableTransactionBuilder`](IotaPTB).
+///
+/// Additional Features:
+/// - Command results can be bound by a static name, and referenced by later
+///   calls with that name using [`Res`].
+/// - Parameters can be passed in without fiddling with inputs, and will be
+///   automatically handled.
+/// - Mixed parameters (new inputs and named references) can be defined in a
+///   single tuple.
+/// - The builder is 100% chainable, start-to-finish.
+#[derive(Debug, Clone)]
+pub struct TransactionBuilder<C> {
+    /// The inputs to the transaction.
+    inputs: Vec<UnresolvedInput>,
+    /// The list of commands in the transaction. A command is a single operation
+    /// in a programmable transaction.
+    commands: Vec<Command>,
+    /// The gas objects that will be used to pay for the transaction. The most
+    /// common way is to use [`Input::owned`] function to create
+    /// a gas object and use the [`add_gas_objects`](Self::add_gas_objects)
+    /// method to set the gas objects.
+    pub(crate) gas: Vec<ObjectReference>,
+    gas_to_resolve: Vec<ObjectId>,
+    /// The gas budget for the transaction.
+    gas_budget: Option<u64>,
+    /// The gas price for the transaction.
+    gas_price: Option<u64>,
+    /// The sender of the transaction.
+    sender: Address,
```

Comment:
I'm wondering if this should also be made optional, because for some dev inspect txs the sender doesn't matter

Reply:
Why require it in `pub fn new(sender: Address) -> TransactionBuilder<()> {` then?

Reply:
But if it's not optional now, this would be a breaking change later. And for dev inspect, one could just use an address like 0x0 as sender as default if not provided

Reply:
Fair point, I just thought it's annoying for users to create an address they don't care about, but I guess it's fine since depending on what they call it can actually have an impact also for dev inspects